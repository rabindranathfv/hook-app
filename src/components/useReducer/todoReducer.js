const ADD = 'add';
const DEL ='delete';
const TOG = 'toggle';
const TOG_OLD = 'toggle-old';

export const todoReducer = ( state = [], action ) => {

    switch ( action.type ) {
        case ADD:
            return [ ...state, action.payload ];
        case DEL:
            return state.filter( todo => todo.id !== action.payload );
        case TOG: 
            return state.map( todo => 
                ( todo.id === action.payload )
                    ? { ...todo, done: !todo.done }
                    : todo
            );
        case TOG_OLD:
            return state.map( todo => {

                if ( todo.id === action.payload ) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                } else {
                    return todo;
                }
            })
        default:
            return state;
    }


}