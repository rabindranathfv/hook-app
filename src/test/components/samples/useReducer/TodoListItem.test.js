import React from 'react';
import { shallow } from 'enzyme';
import { TodoListItem } from '../../../../components/useReducer/TodoListItem';

const todoMock = { id: 1, desc: 'descripcionTest1', done: true };
const handleDeleteMock = jest.fn();
const handleToggleMock = jest.fn();

describe('TodoListItem', () => {

    let wrapper = shallow(<TodoListItem todo={todoMock} index={todoMock.id} handleDelete={handleDeleteMock} handleToggle={handleToggleMock}/> );

    beforeEach(() => {
        wrapper = shallow(<TodoListItem todo={todoMock} index={todoMock.id} handleDelete={handleDeleteMock} handleToggle={handleToggleMock}/> );
    });
  
    it('should render todoListItem', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render todoItem description and index', () => {
        const todoDesc = wrapper.find('p');

        expect(todoDesc.text()).toBe(`${todoMock.id+1}. ${todoMock.desc}`);
    });

    it('should render todo with class complete when its done', () => {
        const completeTodo =  wrapper.find('p');

        expect(completeTodo.hasClass('complete')).toBeTruthy();
    });
    
    it('should render and apply handleToggle', () => {
        const findToggle = wrapper.find('p');
        findToggle.simulate('click');

        expect(typeof findToggle.props().onClick).toBe('function');
        expect(handleToggleMock).toBeCalled();
        expect(handleToggleMock).toHaveBeenCalledWith(1);
    });

    it('should render and apply handleDelete', () => {
        const findDelete = wrapper.find('button');
        findDelete.simulate('click');
        
        expect(findDelete.text()).toBe('DELETE');
        expect(findDelete.props().className).toBe('btn btn-danger');  
        expect(typeof findDelete.props().onClick).toBe('function');
        expect(handleDeleteMock).toBeCalled();
        expect(handleDeleteMock).toHaveBeenCalledWith(1);
    });
});
