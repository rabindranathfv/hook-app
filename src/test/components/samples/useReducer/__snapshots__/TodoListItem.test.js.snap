// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem should render todoListItem 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="complete"
    onClick={[Function]}
  >
    2
    . 
    descripcionTest1
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    DELETE
  </button>
</li>
`;
